File: Examples/Python/Ring.py
import sys
import System
import clr
clr.AddReference("MPI")
import MPI

args = System.Environment.GetCommandLineArgs()
argsref = clr.Reference[System.Array[str]](args)
env = MPI.Environment(argsref)

comm = MPI.Communicator.world
if (comm.Rank == 0):
    # program for rank 0
    comm.Send[str]("Rosie", 1, 0)

    # receive the final message
    msg = comm.Receive[str].Overloads[(int, int)](MPI.Communicator.anySource, 0)

    print "Rank " + str(comm.Rank) + " received message \"" + msg + "\"."

else: # not rank 0
    # program for all other ranks
    msg = comm.Receive[str].Overloads[(int, int)](comm.Rank - 1, 0)

    print "Rank " + str(comm.Rank) + " received message \"" + msg + "\"."

    comm.Send[str](msg + ", " + str(comm.Rank), (comm.Rank + 1) % comm.Size, 0)

env.Dispose()

File: Examples/Ring/Ring.cs
/* Copyright (C) 2007  The Trustees of Indiana University
 *
 * Use, modification and distribution is subject to the Boost Software
 * License, Version 1.0. (See accompanying file LICENSE_1_0.txt or copy at
 * http://www.boost.org/LICENSE_1_0.txt)
 *  
 * Authors: Douglas Gregor
 *          Andrew Lumsdaine
 * 
 * This example program passes a message around a ring of processes,
 * where each processor adds its rank to the message string.
 */
using System;
using System.Collections.Generic;
using System.Text;
using MPI;

class Ring
{
    static void Main(string[] args)
    {
        using (MPI.Environment env = new MPI.Environment(ref args))
        {
            Intracommunicator comm = MPI.Communicator.world;
            if (comm.Size < 2)
            {
                // Our ring needs at least two processes
                System.Console.WriteLine("The Ring example must be run with at least two processes.");
                System.Console.WriteLine("Try: mpiexec -np 4 ring.exe");
            }
            else if (comm.Rank == 0)
            {
                // Rank 0 initiates communication around the ring
                string data = "Hello, World!";

                // Send "Hello, World!" to our right neighbor
                comm.Send(data, (comm.Rank + 1) % comm.Size, 0);

                // Receive data from our left neighbor
                comm.Receive((comm.Rank + comm.Size - 1) % comm.Size, 0, out data);

                // Add our own rank and write the results
                data += " 0";
                System.Console.WriteLine(data);
            }
            else
            {
                // Receive data from our left neighbor
                String data;
                comm.Receive((comm.Rank + comm.Size - 1) % comm.Size, 0, out data);

                // Add our own rank to the data
                data = data + " " + comm.Rank.ToString() + ",";

                // Pass on the intermediate to our right neighbor
                comm.Send(data, (comm.Rank + 1) % comm.Size, 0);
            }
        }
    }
}
